### PR TITLE
Fix for: Bug with specific text and bbcode plugin

### LINKED ISSUE #2782: Bug with specific text and bbcode plugin
## Type of report

Bug

## Provide detailed reproduction steps (if any)

1. Open https://ckeditor.com/docs/ckeditor4/latest/examples/bbcode.html
2. Remove all text
3. Enter [[][/url]
4. Switch to Source View
5. Return to normal view

### Expected result

It should show the previously entered text "[[][/url]".

### Actual result

Empty text field and all controls are disabled. Editor "broken".

## Other details

```
Uncaught TypeError: Cannot read property 'type' of null
    at CKEDITOR.BBCodeParser.e.onTagClose (plugin.js?t=J0EE:12)
    at CKEDITOR.BBCodeParser.parse (plugin.js?t=J0EE:9)
    at Function.CKEDITOR.htmlParser.fragment.fromBBCode (plugin.js?t=J0EE:13)
    at a.b (plugin.js?t=J0EE:17)
    at a.e (ckeditor.js:10)
    at a.<anonymous> (ckeditor.js:11)
    at a.window.CKEDITOR.window.CKEDITOR.dom.CKEDITOR.editor.CKEDITOR.editor.fire (ckeditor.js:13)
    at a.setData (ckeditor.js:271)
    at d (ckeditor.js:1215)
    at Object.wysiwyg (ckeditor.js:1217)
```

### PR DESCRIPTION
**Pull request by engineering-this**
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [X] Unit tests
- [X] Manual tests

## What changes did you make?

Transform square brackets in text into html entities.

Closes #2782

### CODE REVIEW

File: plugins/bbcode/plugin.js [hunk 1]
```diff
@@ -651,7 +665,17 @@
...
 
+			var entities = {
+				'[': '&#91;',
+				']': '&#93;'
+			};
+
 			editor.dataProcessor.htmlFilter.addRules( {
+				text: function( value ) { // Replace square brackets with entities (#2782).
+					return value.replace( /(\[|\])/g, function( match ) {
+						return entities[ match ];
```
Comment (Comandeer) [inline plugins/bbcode/plugin.js]: I'm wondering if we can use `String.prototype.charCodeAt` here and remove `entities` array altogether:

```suggestion
						return '&#' + match.charCodeAt( 0 ) + ';';
```

File: plugins/bbcode/plugin.js [hunk 2]
```diff
@@ -414,6 +414,20 @@
 			currentNode = parent;
 		}
 
+		var entities = {
+				'&#91;': '[',
+				'&#93;': ']'
+			},
+		filter = new CKEDITOR.htmlParser.filter( {
+			text: function( value ) { // Replace entities with square brackets (#2782).
+				return value.replace( /(&amp;#91;|&amp;#93;)/g, function( match ) {
```
Comment (Comandeer) [inline plugins/bbcode/plugin.js]: The same here:

```suggestion
				return value.replace( /&amp;#(91|93);/g, function( match, entity ) {
					return String.fromCharCode( entity );
				} );
```

File: plugins/bbcode/plugin.js [hunk 3]
```diff
@@ -414,6 +414,20 @@
 			currentNode = parent;
 		}
 
+		var entities = {
+				'&#91;': '[',
+				'&#93;': ']'
+			},
+		filter = new CKEDITOR.htmlParser.filter( {
```
Comment (Comandeer) [inline plugins/bbcode/plugin.js]: Wrong level of indentation.

File: tests/plugins/bbcode/manual/squarebrackets.html [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<textarea id="editor"></textarea>
+
+<script>
+	CKEDITOR.replace( 'editor', {
+		removeButtons: 'Strike,Subscript,Superscript'
```
[the remark this answers is not in the document]
Comment (engineering-this) [reply tests/plugins/bbcode/manual/squarebrackets.html]: They are not supported by bbcode plugin.

https://github.com/ckeditor/ckeditor-dev/blob/eada03836613023cf3dbb1a8be30294733fa4e8e/plugins/bbcode/plugin.js#L29-L33

**Comment by engineering-this** [comment]
Using charCode works well, good idea!

**Pull request closed without merging**